### PR TITLE
Fix bug 1957852 - prevent vm to start when restore is in progress

### PR DESCRIPTION
**Pull request by ShellyKa13**
**What this PR does / why we need it**:
When vm restore in progress prevent from starting the vm until the restore is done

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes # https://bugzilla.redhat.com/show_bug.cgi?id=1957852

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go [hunk 1]
```diff
@@ -433,6 +443,30 @@ func (admitter *VMsAdmitter) validateVolumeRequests(vm *v1.VirtualMachine) ([]me
 
 }
 
+func validateRestoreStatus(ar *admissionv1.AdmissionRequest, vm *v1.VirtualMachine) []metav1.StatusCause {
```
Comment (arnongilboa) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go:446]: better refactor this one together with validateSnapshotStatus which are almost similar

File: pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter_test.go [hunk 2]
```diff
@@ -1210,6 +1210,59 @@ var _ = Describe("Validating VM Admitter", func() {
 			return true
 		}),
 	)
+
+	table.DescribeTable("when restore is in progress, should", func(mutateFn func(*v1.VirtualMachine) bool) {
+		vmi := v1.NewMinimalVMI("testvmi")
+		vm := &v1.VirtualMachine{
+			Spec: v1.VirtualMachineSpec{
+				Running: &[]bool{false}[0],
```
Comment (arnongilboa) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter_test.go:1218]: why not `Running: &notRunning` ?

File: pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter_test.go [hunk 3]
```diff
@@ -1210,6 +1210,59 @@ var _ = Describe("Validating VM Admitter", func() {
...
+				Template: &v1.VirtualMachineInstanceTemplateSpec{
+					Spec: vmi.Spec,
+				},
+			},
+			Status: v1.VirtualMachineStatus{
+				RestoreInProgress: &[]string{"testrestore"}[0],
```
Comment (arnongilboa) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter_test.go:1224]: same here, use something like:
```
testRestore := "testrestore"
...
RestoreInProgress: &testRestore
```

File: pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter_test.go [hunk 4]
```diff
@@ -1210,6 +1210,59 @@ var _ = Describe("Validating VM Admitter", func() {
...
+		resp := vmsAdmitter.Admit(ar)
+		Expect(resp.Allowed).To(Equal(allow))
+
+		if !allow {
+			Expect(len(resp.Result.Details.Causes)).To(Equal(1))
+			Expect(resp.Result.Details.Causes[0].Field).To(Equal("spec"))
+		}
+	},
+		table.Entry("reject update to spec", func(vm *v1.VirtualMachine) bool {
+			vm.Spec.Running = &[]bool{true}[0]
```
Comment (arnongilboa) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter_test.go:1254]: same as above

Comment (mhenriks) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter_test.go:1254]: the alternative is more lines of code

**Comment by mhenriks** [comment]
So far so good!  Think we just need a functional test.  To make testing possible may want to use the dummy webhook method similar to here:  https://github.com/kubevirt/kubevirt/blob/master/tests/storage/restore.go#L306

But may have to do on a different resource/operation like "create PVC" or something

File: pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter_test.go (hunk 4 above)
Comment (arnongilboa) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter_test.go:1254]: Whatever, but I think the addititional 1 line will make it cleaner and consistent with the rest of the code...

**Comment by ShellyKa13** [comment]
@mhenriks I fixed some code and added the functional test we talked about. Will appreciate a new review

**Pull request merged**